src/archive/mod.rs

```diff
@@ -139,24 +188,89 @@ pub struct Index {
         Ok (Index {
             size: sizeof_table,
             symbol_indexes: indexes,
             strtab: strtab.to_vec()?, // because i'm lazy
         })
     }
+
+    /// Parses the given byte buffer into an Index, in BSD style archives
+    pub fn parse_bsd_symdef(buffer: &[u8]) -> Result<Index> {
+        // `llvm-ar` is a suitable reference:
+        //   https://github.com/llvm-mirror/llvm/blob/6ea9891f9310510c621be562d1c5cdfcf5575678/lib/Object/Archive.cpp#L842-L870
+
+        // BSD __.SYMDEF files look like:
+        //
+        //            ┌─────────────┐
+        //  entries:  │   # bytes   │
+        //            ├─────────────┼─────────────┐
+        //            │ name offset │  .o offset  │
+        //            ├─────────────┼─────────────┤
+        //            │ name offset │  .o offset  │
+        //            ├─────────────┼─────────────┤
+        //            │ name offset │  .o offset  │
+        //            ├─────────────┼─────────────┤
+        //            │ name offset │  .o offset  │
+        //            ├─────────────┼─────────────┘
+        //   strings: │   # bytes   │
+        //            ├─────────────┴───────────────────┐
+        //            │  _symbol\0                      │
+        //            ├─────────────────────────────────┴─────────────────────┐
+        //            │  _longer_symbol\0                                     │
+        //            ├────────────────┬──────────────────────────────────────┘
+        //            │  _baz\0        │
+        //            ├────────────────┴───┐
+        //            │  _quxx\0           │
+        //            └────────────────────┘
+        //
+        // All numeric values are u32s. Name offsets are relative to the start of the string table,
+        // and .o offsets are relative to the the start of the archive.
+
+        // read the number of entries
+        let entries_bytes = buffer.pread_with::<u32>(0, scroll::LE)? as usize;
+        let entries = entries_bytes / 8;
+
+        // set up the string table
+        let strtab_bytes = buffer.pread_with::<u32>(entries_bytes + 4, scroll::LE)? as usize;
+        let strtab = strtab::Strtab::parse(buffer, entries_bytes + 8, strtab_bytes, 0x0)?;
+
+        // build the index
+        let mut indexes = Vec::with_capacity(entries);
+        let mut strings: Vec<String> = Vec::with_capacity(entries);
+        for i in 0..entries {
+            let string_offset: u32 = buffer.pread_with(i * 8 + 4, scroll::LE)?;
+            let archive_member: u32 =  buffer.pread_with(i * 8 + 8, scroll::LE)?;
+
+            let string = match strtab.get(string_offset as usize) {
+                Some(result) => result,
+                None => Err(Error::Malformed(format!("{} entry {} has string offset {}, which is out of bounds", BSD_SYMDEF_NAME, i, string_offset)))
+            }?;
+
+            indexes.push(archive_member);
+            strings.push(string.to_owned());
```

Comment:
I was going for parity with the existing SysV implementation. I'll switch both to zero copy.